Vcpkg Package Manager

mrixner opened the pull request:
<!-- Provide a general summary of your changes in the title above -->
- [x] **I have read the [contributing guidelines](https://github.com/marticliment/WingetUI/blob/main/CONTRIBUTING.md#coding), and I agree with the [Code of Conduct](https://github.com/marticliment/WingetUI/blob/main/CODE_OF_CONDUCT.md)**.
- [x] **Have you checked that there aren't other open [pull requests](https://github.com/marticliment/WingetUI/pulls) for the same changes?**
- [x] **Have you tested that the committed code can be executed without errors?**
- [x] **This PR is not composed of garbage changes used to farm GitHub activity to enter potential Crypto AirDrops.**
Any user suspected of farming GitHub activity with crypto purposes will get banned. Submitting broken code wastes the contributors' time, who have to spend their free time reviewing, fixing, and testing code that does not even compile breaks other features, or does not introduce any useful changes. I appreciate your understanding.
-----

# TODO before merging (for @marticliment)
 -  [x] Add the VCPkg icon to symbol icon table
 -  [x] Add the setting to specify the default triplet.


-----


This pull request is for support for the vcpkg package manager.

I'm sorry if the code is not good, this is my first time ever working with C#. I tried to write code in the style of all the other package managers and not touch more than I needed to.

## List of Added Translations:
- vcpkg ManagerProperties Description (A popular C/C++ library manager. Full of C/C++ libraries and other C/C++-related utilities&lt;br&gt;Contains: &lt;b&gt;C/C++ libraries and related utilities&lt;/b&gt;)
- InvalidOperationException value (Cannot update vcpkg port files as requested: git was not installed or the VCPKG_ROOT environment variable / the custom vcpkg root setting were not set)
- Automatic Portfile Update Setting Title (Update vcpkg's Git portfiles automatically (requires Git installed))
- Default vcpkg Triplet Setting Title (Default vcpkg triplet)
- Custom vcpkg Root Setting Title (Custom vcpkg root)

## Manager Capabilities (for the table on the home page):

**Install as administrator**: yes (for vcpkg roots in an admin-locked directory)
**Skip integrity checks**: no (I don't think vcpkg has this capability, or if it has integrity checks to skip)
**Interactive installation**: no (it's being built from the command-line; there's no interactivity to show)
**Install Older Versions**: no (vcpkg doesn't support this on a global scale, only for projects with [vcpkg.json](https://learn.microsoft.com/en-us/vcpkg/reference/vcpkg-json) control files)
**Install a PreRelease Version**: no (vcpkg doesn't have pre-release versions)
**Install a Custom Architecture**: not directly supported but can be easily achieved (via using a different triplet / source)
**Install on a Custom Scope**: no (everything just gets installed to the vcpkg directory)
**Custom Install Location**: no / sort of not directly supported but can be easily achieved? (everything just gets installed to the vcpkg directory; you can change which vcpkg directory via the "Custom vcpkg root" setting)
**Custom Package Sources**: not directly supported but can be easily achieved (it will automatically detect custom triplet / source CMake scripts dropped into `%VCPKG_ROOT%\triplets` or `%VCPKG_ROOT%\triplets\community`)

## Other Potentially Unresolved Things

- I have added a colored icon, but as [src/UniGetUI.PackageEngine.Managers.Vcpkg/Vcpkg.cs:65](https://github.com/mrixner/UniGetUI/tree/vcpkg-package-manager/src/UniGetUI.PackageEngine.Managers.Vcpkg/Vcpkg.cs) states, discussion #2826 brought me to believe that an outlined icon must be added by the maintainers, so if you care enough that may be one thing that needs to happen.
- I would recommend probably squash-merging this if you weren't going to already, as I pretty much just edited the file for package manager-specific settings and the files for the package manager. Unfortunately, I accidentally forgot to commit some changes for some commits, so not all commits compile as requested in the Contributing Guidelines.
- As mentioned in #2146: The `DefaultVcpkgTriplet` setting isn't being set properly, but I was told to leave the UI to Marti, and so that's what I'm doing.

-----
<!-- insert below the issue number (if applicable, do not create a new issue to justify a PR) -->
Closes #2146

marticliment commented:
Hey, I have been looking at the code, and I have found this:
![image](https://github.com/user-attachments/assets/f06e2cf8-95a7-4ffc-bc1e-5c97faefaaa9)

There is an overridable method on the PackageManager class (`public void RefreshPackageIndexes()`) that will be called by UniGetUI when needed to update the sources. 
I will move the code there, but I reckon if you then think the `Settings.Get("UpdateVcpkgGitPorts")` is still needed or not

marticliment commented:
Also, do you think it would make sense to list `git` as a dependency for vcpkg, so if the user doesn't have git installed UniGetUI prompts them to install it?

marticliment commented:
File: VcpkgSourceProvider.cs
![image](https://github.com/user-attachments/assets/02218ae5-5215-4291-9cee-e68c7585b126)


What is the idea here?
GetAddSourceParameters and GetRemoveSourceParameters are supposed to return the command-line paramaters to add or remove a source, which I understand does not make sense when talking about triplets. Therefore, `_getAddSourceOperationVeredict` and `_getRemoveSourceOperationVeredict` would be never called, right?

marticliment commented:
Please try the changes I have pushed to confirm they work for you.
When we resolve the questions/comments mentioned above, I will merge the PR

mrixner commented:
Sorry for the late response (and for making you read all that code...).

> There is an overridable method on the PackageManager class (`public void RefreshPackageIndexes()`) that will be called by UniGetUI when needed to update the sources.

Sorry, I didn't realize that.

> I reckon if you then think the `Settings.Get("UpdateVcpkgGitPorts")` is still needed or not

I'm not sure. I guess it depends on how much control you want to give the user; I assumed not everyone would want the vcpkg repository getting automatically updated, but I suppose you're right; it doesn't matter.

I see you changed the setting to `DisableUpdateVcpkgGitPorts`; I think that makes more sense. I feel like you should be able to say "no, don't automatically update my git repository", in case maybe you're working in it or have it at a certain commit for some reason? But updates should be enabled by default. I think the current implementation with `DisableUpdateVcpkgGitPorts` is fine, unless you don't want the setting, as it's not strictly necessary to the operation of vcpkg.

> Also, do you think it would make sense to list `git` as a dependency for vcpkg, so if the user doesn't have git installed UniGetUI prompts them to install it?

I don't think so. Aside from the fact that people using vcpkg probably already have git installed, given that you need to have cloned the git repository in order for vcpkg to work properly git will probably be installed if they have vcpkg. Although - it won't hurt anything and is probably better for clarity, so maybe.

> What is the idea here?
> GetAddSourceParameters and GetRemoveSourceParameters are supposed to return the command-line paramaters to add or remove a source, which I understand does not make sense when talking about triplets. Therefore, _getAddSourceOperationVeredict and _getRemoveSourceOperationVeredict would be never called, right?

I don't know if they're called; I'm not sure how source management works. I assumed they would be called anyway when a source was added in the settings, and so the code there will return whether or not a given source is valid (recognized by vcpkg).

I tried removing the custom sources ability, since the code should autodetect any added triplets, but it removed the triplet sorting in the Available Updates section which I thought should be there, but you're right; those four functions don't _really_ serve a purpose.

I see you removed its ability for custom sources. This doesn't change any of the actual functionality; again, the only reason _I_ didn't do this was for the by-triplet selection in the available updates list.

> Please try the changes I have pushed to confirm they work for you.
> When we resolve the questions/comments mentioned above, I will merge the PR

I'll look at the changes and see if it works. I'll also change some of the other things I found.

mrixner commented:
OK, looking through the UI, the settings UIs look really great, and everything major works as intended. However, there's a few things I noticed.

- The default triplet dropdown shows a restart required banner when it doesn't need to (fixed in 2e5a166)
- I was hoping the first time you viewed the default triplet dropdown the value would be the default triplet as determined elsewhere, however this is a tiny nitpick and I understand if it's not possible.
- In the "discover" list of searched packages, if you have a package installed for a different triplet than the one you're searching for it still shows the little "installed" icon. I'm not sure what it is using to compare that, but I assumed that given that the package IDs are different, that would be different as well?
- As mentioned in the last comment, due to removal of custom sources, you can no longer filter the vcpkg packages by their triplet. This is also trivial and possibly intended / doesn't matter, but if there's an easy way to add that I think that could be helpful.

I'll look through all the diffs to see if there's anything else I missed.

mrixner commented:
I removed the unnecessary restart required banner, fixed a typo, and made `Vcpkg.GetDefaultTriplet` public. I did not remove any of the TODOs dealing with `DisableUpdateVcpkgGitPorts`, in case it's not something you want to keep.

I noticed this in the commit log; is there an implementation to think about? (it's been shifted to SettingPage`.xaml.cs:441`)
![image](https://github.com/user-attachments/assets/f2ba68e0-8275-430d-8d6e-82b2627f52f3)

marticliment commented:
> * As mentioned in the last comment, due to removal of custom sources, you can no longer filter the vcpkg packages by their triplet. This is also trivial and possibly intended / doesn't matter, but if there's an easy way to add that I think that could be helpful.

Oh, this actually makes sense. 
What I will do is I will add source compatibility again, but I will manually remove the Source Manager UI from the settings window.



> * I was hoping the first time you viewed the default triplet dropdown the value would be the default triplet as determined elsewhere, however this is a tiny nitpick and I understand if it's not possible.

This can be _kinda_ fixed. On the LoadManager() method, you can check if the setting is empty, and if it is not set, it can be set to a default value. I don't know how to get the system default triplet, but if you know how to and you want to fix this it'd be great, 


> * In the "discover" list of searched packages, if you have a package installed for a different triplet than the one you're searching for it still shows the little "installed" icon. I'm not sure what it is using to compare that, but I assumed that given that the package IDs are different, that would be different as well?

If I enable source support again this issue should fix itself. 

> I noticed this in the commit log; is there an implementation to think about? (it's been shifted to SettingPage`.xaml.cs:441`)

This might me my fault, VS might have created it automatically and then I used a lambda. I will get this fixed.

mrixner commented:
> What I will do is I will add source compatibility again, but I will manually remove the Source Manager UI from the settings window.

Sounds good, thank you!

> This can be kinda fixed. On the LoadManager() method, you can check if the setting is empty, and if it is not set, it can be set to a default value. I don't know how to get the system default triplet, but if you know how to and you want to fix this it'd be great,

Yes, I noticed and realized this as I was looking through the settings to add a restart required banner for #2883. I'll go do that in a second.

> If I enable source support again this issue should fix itself.

Ahh, of course. That makes sense.

mrixner commented:
> If I enable source support again this issue should fix itself.

Unfortunately, it doesn't seem to have fixed itself.

![image](https://github.com/user-attachments/assets/76db5960-154e-4d94-a53c-b787c9832cbe)
![image](https://github.com/user-attachments/assets/a9421801-4319-4025-a986-4456970304ad)

Everything else is looking good, though!

marticliment commented:
> **CORRECTION**: this doesn't work when an item was just uninstalled, but it does if it's not installed when UniGetUI starts. Therefore, this is not an issue.

This is a bug I am aware of, uninstalled packages still show as installed on the discover page

mrixner commented on src/UniGetUI.PackageEngine.Managers.Pip/Pip.cs:
```diff
@@ -18,6 +19,17 @@ public class Pip : PackageManager
         public Pip()
         {
+            Dependencies = [
+                // GIT is required for vcpkg updates to work
```
This comment annotates this wrong; shouldn't it be in reference to Pip instead?

marticliment commented:
I will merge once I add the icon. The rest looks fine for me

mrixner commented:
> I will merge once I add the icon. The rest looks fine for me

OK, thank you! Do you need me to make the icon, or will you do that yourself?

mrixner commented:
Also, just a note in case any issues are opened about vcpkg package discovery taking a while - I have tested this by running `vcpkg search` from the command line, and the first `search` / `list` operation every once in a while just takes a rather long time; this isn't an issue with UniGetUI.

marticliment commented:
> OK, thank you! Do you need me to make the icon, or will you do that yourself?

I'll do this myself

marticliment commented:
Vcpkg icon
![image](https://github.com/user-attachments/assets/1818cd24-b682-42fb-86e0-c8ec334132ec)
![image](https://github.com/user-attachments/assets/7f4dae94-09c2-400a-8b51-73df654a2a8b)

mrixner commented:
> Vcpkg icon ![image](https://private-user-images.githubusercontent.com/53119851/384729608-1818cd24-b682-42fb-86e0-c8ec334132ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEyNzcwNjAsIm5iZiI6MTczMTI3Njc2MCwicGF0aCI6Ii81MzExOTg1MS8zODQ3Mjk2MDgtMTgxOGNkMjQtYjY4Mi00MmZiLTg2ZTAtYzhlYzMzNDEzMmVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEwVDIyMTI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhMzMwMjgwMmIxNGE4MTBlMGE2OTQwZTY1N2VhMDY2OTg3YmU5ODVjMTc3YmVjZTg5MTgyMDI5YWVkZDcwMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.12bqqdb0WEEgmcXSqCt9bB3I82OVkZsMWcqFltK5AN0) ![image](https://private-user-images.githubusercontent.com/53119851/384729617-7f4dae94-09c2-400a-8b51-73df654a2a8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEyNzcwNjAsIm5iZiI6MTczMTI3Njc2MCwicGF0aCI6Ii81MzExOTg1MS8zODQ3Mjk2MTctN2Y0ZGFlOTQtMDljMi00MDBhLThiNTEtNzNkZjY1NGEyYThiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEwVDIyMTI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1MmEwOWQxMTc5MDkyMTBiOGNjOWI1ZTY4ODA2YWEwMDI2NGQ3NzI4NjRlMDFjNDQ3Mjc5YTJkZDk3YzZiYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bEDR1jfUM0HVVat80u8jhstbEmP6Nm1dEMC3KyYFK4k)

Looks great!
